src/adapter/src/catalog/state.rs

```diff
@@ -1767,13 +1776,40 @@ impl CatalogState {
             },
         };
 
-        for (database_spec, schema_spec) in schemas {
-            let schema = self.get_schema(&database_spec, &schema_spec, conn_id);
+        for (database_spec, schema_spec) in &schemas {
+            let schema = self.get_schema(database_spec, schema_spec, conn_id);
 
             if let Some(id) = get_schema_entries(schema).get(&name.item) {
                 return Ok(&self.entry_by_id[id]);
             }
         }
+
+        // Some relations that have previously lived in the `mz_internal` schema have been moved to
+        // `mz_catalog_unstable` or `mz_introspection`. To simplify the transition for users, we
+        // automatically let uses of the old schema resolve to the new ones as well.
+        // TODO(#27831) remove this after sufficient time has passed
+        let mz_internal_schema = SchemaSpecifier::Id(self.get_mz_internal_schema_id());
+        if schemas.iter().any(|(_, spec)| *spec == mz_internal_schema) {
+            for schema_id in [
+                self.get_mz_catalog_unstable_schema_id(),
+                self.get_mz_introspection_schema_id(),
+            ] {
+                let schema = self.get_schema(
+                    &ResolvedDatabaseSpecifier::Ambient,
+                    &SchemaSpecifier::Id(schema_id),
+                    conn_id,
+                );
+
+                if let Some(id) = get_schema_entries(schema).get(&name.item) {
+                    debug!(
```

Comment:
Could this be a metric instead of a debug that defaults to off? As a metric we'd be able to easily have certainty when we could safely remove this.

Reply:
Metrics are unfortunately not available in this context (`CatalogState`). We can add them, but I'd rather to that in a different PR to keep the scope of this one small.